### PR TITLE
Work

### PR DESCRIPTION
**Pull request by rdube**
@foehnph @Messikommer for your info! :-)

### CODE REVIEW

File: src/attitude_filter.cpp [hunk 1]
```diff
@@ -9,15 +9,22 @@
 #include "geometry_msgs/Quaternion.h"
 #include "sensor_msgs/Imu.h"
 
+//Probleme:
+// 1. typedef--> error: not declared in scope
+// 2. Implementation von Methoden einer Klasse in c++ File ?
+// 3. KalmanFilter_Orientation kalman; while ... kalman() did not, why ??
+
+//typedef 
+// typedef Eigen::MatrixXd::Identity(3) Identity_3;
```
Comment (rdube) [inline src/attitude_filter.cpp]: rm dead code

File: src/attitude_filter.cpp [hunk 2]
```diff
@@ -28,58 +35,46 @@ const float error_measurement_gyro = 0.001;
...
 //Initialization for kalman Update (global: no new initialization every step)
 Eigen::Vector3d current_angles;
 Eigen::VectorXd current_state;
-Eigen::VectorXd current_measurements;
 geometry_msgs::Quaternion current_orientation;
 
-//Initialising Kalman Filter
-KalmanFilter kalman(A, B, C, Q, R);
+//Declaration Kalman Filter
+KalmanFilter_Orientation kalman;
```
Comment (rdube) [inline src/attitude_filter.cpp]: No underscore in class names. Why only orientation? You want to estimate both position and orientation right?

File: src/attitude_filter.cpp [hunk 3]
```diff
@@ -28,58 +35,46 @@ const float error_measurement_gyro = 0.001;
...
 
 int main(int argc, char** argv){
 
 	ros::init(argc, argv, "attitudefilter");
 	ros::NodeHandle node;
 
+	//Initialising state matrices
+	Eigen::Matrix<double,15,15> A = Eigen::MatrixXd::Identity(15,15);
```
Comment (rdube) [inline src/attitude_filter.cpp]: I believe this is not your latest code as I do not see the class members here. You moved A to a private member in the class right?

**Comment by rdube** [comment]
LGTM (looks good to me) Keep us updated with the latest status. Cheers!

File: src/attitude_filter.cpp (hunk 3 above)
Comment (foehnx) [reply src/attitude_filter.cpp]: Sorry, I ment to comment, rdube's comment xD
I guess he sets up the matrices and passes it to the kalman in its `init()` method down at row 77.
I would rather recommend doing so in the constructor of the kalman class if possible.
@SeleSchaefer google for "c++ member initializer list", it's the cleanest way to do this.

File: src/attitude_filter.cpp [hunk 4]
```diff
@@ -28,58 +35,46 @@ const float error_measurement_gyro = 0.001;
...
+	//Initialising inital state
+	const Eigen::VectorXd x0 = Eigen::VectorXd::Zero(9,9);
+	const Eigen::Vector3d init_angles(x0(3), x0(4), x0(5));
 
 	//Initialising Kalman Filter
-	kalman.init(x0);
+	kalman.init(x0,A,B,C,Q,R);
 
 	//Subscribing & Update loop
 	while(ros::ok()){
```
Comment (foehnx) [inline src/attitude_filter.cpp]: propably you wanna use just `ros.spin()` instead of a while loop.

File: src/attitude_filter.cpp [hunk 5]
```diff
@@ -95,44 +90,26 @@ int main(int argc, char** argv){
...
+	kalman.update(dt, imu_data);
 
 	//Transforming into quaternions
 	current_state = kalman.state();
-	current_angles(0) = current_state(4);
-	current_angles(1) = current_state(5);
-	current_angles(2) = current_state(6);
+	current_angles(0) = current_state(3);
+	current_angles(1) = current_state(4);
+	current_angles(2) = current_state(5);
```
Comment (foehnx) [inline src/attitude_filter.cpp]: assign using `.segment()` or a similar one-liner, thats the nice part of Eigen.

**Comment by rdube** [comment]
@SeleSchaefer nice work with parameters! Could you test and load them correctly?

**Comment by rdube** [comment]
That code is looking good! :)

**Pull request merged**